### PR TITLE
Firebase link in for Android

### PR DESCRIPTION
**Pull request by OSPFNeighbour**
Can someone else test that the app still builds for them.

Had pretty big changes to get this working:

Android Only:

Upgraded Gradle to Version 4.4 (induced a bunch of depreciation warnings in our included libs)
Change gradle config where possible to the new 'implementation' methods
Upgraded Google build tools to 3.1.0
Updated Google services to 3.2.0
Upgraded Google Play Services to 12.0.1
Added permission to vibrate (as recommended by FCM)
Added Maven repo
Linked in Firebase Core + Firebase Message + Firebase Notifications + ShortcutBadger

Message and Notifications seems to go hand in hand so ill throw them both in. Cant seem to get the registration key for the device without the message code in there.... so ill keep messages incase we want to play with that later.

So this was totes fun.

### CODE REVIEW

File: client/android/app/build.gradle [hunk 1]
```diff
@@ -149,25 +149,36 @@ android {
 }
 
 dependencies {
-    compile project(':bugsnag-react-native')
-    compile project(':react-native-device-info')
-    compile (project(':react-native-maps')){
+    implementation project(':bugsnag-react-native')
```
Comment (sdunster) [inline client/android/app/build.gradle]: do you know if these changes mean we can't use `react-native link` anymore? You could test by finding a native module and trying to link it (make sure you do it with a clean repo so you can wipe it afterwards)

File: client/android/build.gradle [hunk 2]
```diff
@@ -2,16 +2,17 @@
...
 buildscript {
     repositories {
         jcenter()
+        google()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.2.3'
-        classpath 'com.google.gms:google-services:3.1.2'
+        classpath 'com.android.tools.build:gradle:3.1.0'
+        classpath 'com.google.gms:google-services:3.2.0'
```
Comment (sdunster) [inline client/android/build.gradle]: i thought we would be including firebase stuff here, no?

File: client/android/app/src/main/java/com/callout/MainApplication.java [hunk 3]
```diff
@@ -38,7 +41,10 @@ public boolean getUseDeveloperSupport() {
             new RNDeviceInfo(),
             new MapsPackage(),
             new CodePush(getResources().getString(R.string.reactNativeCodePush_androidDeploymentKey), getApplicationContext(), BuildConfig.DEBUG),
-            new VectorIconsPackage()
+            new VectorIconsPackage(),
+            new RNFirebasePackage(),
+            new RNFirebaseMessagingPackage(),
```
Comment (sdunster) [inline client/android/app/src/main/java/com/callout/MainApplication.java]: do we need both messaging and notifications? what is the difference?

File: client/package-lock.json [hunk 4]
```diff
@@ -85,6 +251,70 @@
...
     },
+    "onetime": {
+      "version": "2.0.1",
+      "resolved": "https://registry.npmjs.org/onetime/-/onetime-2.0.1.tgz",
+      "integrity": "sha1-BnQoIw/WdEOyeUsiu6UotoZ5YtQ=",
+      "requires": {
+        "mimic-fn": "1.2.0"
+      }
+    },
+    "opencollective": {
```
Comment (sdunster) [inline client/package-lock.json]: :/

File: client/android/app/build.gradle (hunk 1 above)
Comment (OSPFNeighbour) [reply client/android/app/build.gradle]: Both will work, you will just get a "deprecated end of 2018" warning. I assume react will catch up and sort them self out before then

File: client/android/build.gradle (hunk 2 above)
Comment (OSPFNeighbour) [reply client/android/build.gradle]: // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files

Comment (OSPFNeighbour) [reply client/android/build.gradle]: These are build deps, not application deps... *i think*

File: client/android/app/src/main/java/com/callout/MainApplication.java (hunk 3 above)
Comment (OSPFNeighbour) [reply client/android/app/src/main/java/com/callout/MainApplication.java]: messaging is data only and handled by the end app, notifications are key pairs that are displayed by the FCM engine (as we discussed APN are handed internally so lets keep the approach the same)

Comment (OSPFNeighbour) [reply client/android/app/src/main/java/com/callout/MainApplication.java]: also notification doesnt have the functions to fetch a token from the server... weirdly

File: client/package-lock.json (hunk 4 above)
Comment (OSPFNeighbour) [reply client/package-lock.json]: I know. we can fork and remove it if you want

**Comment by sdunster** [comment]
This is way stale and I'm pretty sure we've got everything we need on `sdunster.push_firebase`.

**Comment by sdunster** [comment]
Superseded by #344

**Pull request closed without merging**